Add option for tab complete to use displaynames

cyberkitsune opened the pull request:
<!--

EssentialsX feature submission guide
====================================

NOTE: Failure to fill out this template properly may result in your PR being
      delayed or ignored without warning.

NOTE: Don't type between any arrows in the template, as this text will be
      hidden. This includes this header block and any other explanation text
      blocks.

Want to discuss your PR before submitting it? Join the EssentialsX Development
server: https://discord.gg/CUN7qVb


EssentialsX is GPL
------------------

By contributing to EssentialsX, you agree to license your code under the
GNU General Public License version 3, which can be found at the link below:
https://github.com/EssentialsX/Essentials/blob/2.x/LICENSE


Instructions
------------

If you are submitting a new feature, please follow the following steps:

1.  Fill out the template in full.
      This includes providing screenshots and a link to the original feature 
      request. If there isn't an existing feature request, we strongly
      recommend opening a new feature request BEFORE opening your PR to
      implement it, as this allows us to review whether we're likely to accept
      your feature in advance, and also allows us to discuss possible
      implementations for the feature. If there is no associated feature
      request, your PR may be delayed or rejected without warning.
      
      You can open a new feature request by following this link:
      https://github.com/EssentialsX/Essentials/issues/new/choose

2.  If you are fixing a performance issue, please use the "Bug fix" PR template
      instead. The "bug fix" template is better suited to performance issues.

3.  Include a demonstration.
      If you are adding commands, please provide screenshots and/or a video
      demonstration of the feature. Similarly, if you are adding a new API,
      please include a link to example code that takes advantage of your
      proposed API. This will aid us in reviewing PRs and speed up the process
      significantly.

-->

### Information

<!--
    Replace #nnnn with the number of the original issue. If this PR implements
    features from multiple issues, you should repeat the phrase "closes #nnnn"
    for each issue. 
-->

This PR closes #4431. 

### Details

**Proposed feature:**  
This feature adds a new configuration option, `change-tab-complete-name`

When `change-tab-complete-name` is turned on, `getPlayers()` in `EssentialsCommand` will use `getDisplayName()` instead of `getName()`; Populating the list with display names instead of player names.


**Environments tested:**    

<!-- Type the OS you have used below. -->
OS: Linux, Pterodactyl Java 11 Docker Container

<!-- Type the JDK version (from java -version) you have used below. -->
Java version:  openjdk version "11.0.10" 2021-01-19

<!--
    Put an "x" inside the boxes for the server software you have tested this 
    bug fix on. If this feature does not apply to a server, strike through the server software using ~~strikethrough~~. If you have tested on other
    environments, add a new line with relevant details.
-->
- [X] Most recent Paper version (1.17.1, git-Paper-151)
- [X] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8


**Demonstration:**  
![image](https://user-images.githubusercontent.com/920550/127929360-03ffc600-9dd0-4b0a-a9ee-183d0a09989d.png)

cyberkitsune committed:
Add setting to use a player's display name to ISettings and Settings

cyberkitsune committed:
Use display name when configured on tab complete

cyberkitsune committed:
Strip color codes from display names for better compatibility.

In most cases including the color codes causes "Illegal characters in chat" errors.

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/EssentialsCommand.java:
```diff
@@ -240,7 +240,7 @@ boolean canInteractWith(final CommandSource interactor, final User interactee) {
         for (final User user : ess.getOnlineUsers()) {
             if (canInteractWith(interactor, user)) {
-                players.add(user.getName());
+                players.add(ess.getSettings().changeTabCompleteName() ? ChatColor.stripColor(user.getDisplayName()) : user.getName());
```
Please use `FormatUtil#stripFormat`

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/commands/EssentialsCommand.java:
```diff
@@ -226,7 +226,7 @@ boolean canInteractWith(final CommandSource interactor, final User interactee) {
         for (final User user : ess.getOnlineUsers()) {
             if (canInteractWith(interactor, user)) {
-                players.add(user.getName());
+                players.add(ess.getSettings().changeTabCompleteName() ? ChatColor.stripColor(user.getDisplayName()) : user.getName());
```
Please use `FormatUtil#stripFormat`

JRoy reviewed (changes requested):
A few things plus you should return `Collections.emptyList()` in the following places:

https://github.com/EssentialsX/Essentials/blob/b84207f95510ee6cf0fcc47138771cabe93a1f31/Essentials/src/main/java/com/earth2me/essentials/commands/Commandmsg.java#L63

https://github.com/EssentialsX/Essentials/blob/b84207f95510ee6cf0fcc47138771cabe93a1f31/Essentials/src/main/java/com/earth2me/essentials/commands/Commandmail.java#L298

cyberkitsune committed:
Use `FormatUtil.stripFormat` where possible

cyberkitsune committed:
Send empty list instead of vanilla tab complete in certain situations.

for `/mail` when a recipient is specified
for `/msg` when a recipient is specified

cyberkitsune committed:
this can be simplified now

mdcfe commented on Essentials/src/main/resources/config.yml:
```diff
@@ -49,6 +49,9 @@ hide-displayname-in-vanish: true
 # Disable this if you have any other plugin, that modifies the displayname of a user.
 change-displayname: true
 
+# This option will cause Essentials to display a player's displayname instead of username when tab completing Essentials commands.
```
```suggestion
# This option will cause Essentials to show players' displaynames instead of usernames when tab completing Essentials commands.
```

mdcfe committed:
Update Essentials/src/main/resources/config.yml

mdcfe committed:
Merge branch '2.x' into tabcomplete-displayname
